redis/src/main/scala/zio/redis/options/Connection.scala

```diff
@@ -82,11 +148,269 @@ trait Connection {
+
+  object ClientInfo {
+    implicit val addressSchema: Schema[Option[Address]] = Schema.CaseClass1[String, Option[Address]](
+      TypeId.fromTypeName("Address"),
+      field = Schema.Field[String]("ip", Schema.primitive[String]),
+      construct = ip => {
+        if (ip.contains(":")) {
+          val parts   = ip.split(":")
+          val addr    = parts(0)
+          val portStr = parts(1)
+          Some(Address(InetAddress.getByName(addr), portStr.toInt))
+        } else {
+          None
+        }
+      },
+      extractField = address => address.map(c => c.ip.getHostAddress + ":" + c.port).getOrElse("")
+    )
+    implicit val clientEventsSchema: Schema[ClientEvents] = Schema
+      .CaseClass2(
+        TypeId.fromTypeName("ClientEvents"),
+        field1 = Schema.Field[String]("readable", Schema.primitive[String]),
+        field2 = Schema.Field[String]("writable", Schema.primitive[String]),
+        construct = (readable: String, writable: String) => ClientEvents(readable == "r", writable == "w"),
+        extractField1 = c => if (c.readable) "r" else "",
+        extractField2 = c => if (c.writable) "w" else ""
+      )
+    implicit val clientFlagSchema: Schema[ClientFlag] = Schema
+      .Primitive(StandardType.StringType)
+      .transform[ClientFlag](
+        s => ClientFlag.apply(s.head),
+        _.stringify
+      )
+    implicit val clientFlagsSchema: Schema[Set[ClientFlag]] = Schema
+      .Primitive(StandardType.StringType)
+      .transform[Set[ClientFlag]](
+        s => s.toCharArray.map(ClientFlag.apply).toSet,
+        s => s.map(_.stringify).mkString("")
+      )
+    implicit val schema: Schema[ClientInfo] = DeriveSchema.gen[ClientInfo]
+
+    private val decoder = ClientInfoCodec.decode(schema)
+    private val encoder = ClientInfoCodec.encode(schema)
+
+    def decode(in: String): Either[String, ClientInfo] =
+      decoder(Chunk.fromArray(in.getBytes))
+    def encode(info: ClientInfo): String = new String(encoder(info).toArray)
+  }
+
+  sealed trait ClientListType { self =>
+    private[redis] final def stringify: String =
+      self match {
+        case ClientListType.Normal  => "NORMAL"
+        case ClientListType.Master  => "MASTER"
+        case ClientListType.Replica => "REPLICA"
+        case ClientListType.PubSub  => "PUBSUB"
+      }
+  }
+
+  object ClientListType {
+    case object Normal  extends ClientListType
+    case object Master  extends ClientListType
+    case object Replica extends ClientListType
+    case object PubSub  extends ClientListType
+  }
+
+  sealed trait ClientListFilter {
+    private[redis] def stringify: String
+  }
+
+  object ClientListFilter {
+    case object All extends ClientListFilter {
+      private[redis] def stringify: String = "all"
+    }
+
+    sealed case class Type(clientType: ClientListType) extends ClientListFilter {
+      private[redis] final def stringify: String = s"TYPE"
+    }
+
+    sealed case class Id(ids: Long*) extends ClientListFilter {
```

Comment:
Should this be a non-empty list?

Reply:
Changed to zio.prelude.NonEmptyList[Long]

Reply:
Please don't do that. We encode non empty param lists in a different manner.

Reply:
See [this](https://github.com/zio/zio-redis/blob/master/redis/src/main/scala/zio/redis/Input.scala#L273) and its usages.

Reply:
Ok, I think I already have it, variable ids are encoded now using
```scala
case class Id(id: Long, ids: Long*)
```